File: src/main/java/dev/aparikh/jsonplaceholder/model/ApiResponse.java
package dev.aparikh.jsonplaceholder.model;

/**
 * Generic wrapper class for API responses.
 * @param <T> The type of data contained in the response
 */
public class ApiResponse<T> {
    private String status;
    private T data;
    private String message;

    // Default constructor
    public ApiResponse() {
    }

    // Constructor with all fields
    public ApiResponse(String status, T data, String message) {
        this.status = status;
        this.data = data;
        this.message = message;
    }

    // Static factory methods for common response types
    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>("success", data, null);
    }

    public static <T> ApiResponse<T> error(String message) {
        return new ApiResponse<>("error", null, message);
    }

    // Getters and setters
    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }
}

File: src/main/kotlin/dev/aparikh/jsonplaceholder/model/ApiResponse.kt
package dev.aparikh.jsonplaceholder.model

/**
 * Generic wrapper class for API responses.
 * @param T The type of data contained in the response
 */
data class ApiResponse<T>(
    var status: String? = null,
    var data: T? = null,
    var message: String? = null
) {
    companion object {
        // Static factory methods for common response types
        @JvmStatic
        fun <T> success(data: T): ApiResponse<T> {
            return ApiResponse("success", data, null)
        }

        @JvmStatic
        fun <T> error(message: String): ApiResponse<T> {
            return ApiResponse("error", null, message)
        }
    }
}